File: vepserver/src/main/scala/vep/model/session/SessionSearch.scala
package vep.model.session

import org.joda.time.DateTime
import vep.model.common.{ErrorCodes, VerifiableMultiple}
import vep.utils.DateUtils

case class SessionSearch(t: Option[String], s: Option[String], sd: Option[String], ed: Option[String], o: Option[String], p: Option[Int]) extends VerifiableMultiple {
  def theater = t

  def show = s

  def startDate = sd

  def endDate = ed

  lazy val dtStartDate = startDate.map(DateUtils.toDateTime)
  lazy val dtEndDate = endDate.map(DateUtils.toDateTime)

  def order = o

  def page = p

  override protected def doVerify(): Unit = {
    if (page exists (_ <= 0)) {
      addError("p", ErrorCodes.negativeOrNull)
    }

    if (order exists (!SessionSearchColumns.validOrderColumns.contains(_))) {
      addError("o", ErrorCodes.unknownOrder)
    }

    if (startDate exists DateUtils.isNotIsoDate) {
      addError("sd", ErrorCodes.invalidDate)
    }

    if (endDate exists DateUtils.isNotIsoDate) {
      addError("ed", ErrorCodes.invalidDate)
    }
  }
}

object SessionSearchColumns {
  lazy val orderColumn = Map(
    "default" -> "date",
    "d" -> "date",
    "t" -> "theaterName",
    "s" -> "shows"
  )
  lazy val validOrderColumns = orderColumn.keySet
}

case class SessionSearchParsed(theater: String, theaterName: String, id: Int, canonical: String, date: DateTime)

case class SessionSearchShow(canonical: String, title: String)

case class SessionSearchResult(canonical: String, theater: String, theaterName: String, shows: Seq[SessionSearchShow], date: DateTime)

case class SessionSearchResponse(sessions: Seq[SessionSearchResult], pageMax: Int)

File: vepweb/lib/model/session/SessionSearch.dart
part of vep.model.session;

class SessionSearchCriteria extends ModelToString {
  String theater;
  String show;
  DateTime startDate;
  DateTime endDate;
  String order;
  int page;

  void consider(Map<String, Object> filters) {
    theater = filters.containsKey('theater') ? filters['theater'] : theater;
    show = filters.containsKey('shows') ? filters['shows'] : show;
    page = filters.containsKey('page') ? filters['page'] : page;
    if (filters.containsKey('date')) {
      var dates = filters['date'];
      if (dates != null && dates is List<DateTime> && (dates as List<DateTime>).length >= 2) {
        startDate = dates[0];
        endDate = dates[1];
      }
    }
  }
}

class SessionSearchResponse {
  int pageMax;
  List<SessionSearchItem> sessions;
}

class SessionSearchItem {
  String canonical;
  String theater;
  String theaterName;
  List<SessionSearchShow> showList;
  DateTime date;

  String get shows => showList != null ? showList.map((_) => _ != null ? _.title : null).where((_) => _ != null).join(', ') : '';
}

class SessionSearchShow {
  String canonical;
  String title;
}

void prepareJsonxSessionSearch() {
  jsonx.jsonToObjects[SessionSearchResponse] = (json) {
    var sessionSearchResponse = new SessionSearchResponse();
    sessionSearchResponse.pageMax = json['pageMax'];
    sessionSearchResponse.sessions = (json['sessions'] as List<Map<String, Object>>).map((sessionJson) {
      var session = new SessionSearchItem();
      session.canonical = sessionJson['canonical'];
      session.theater = sessionJson['theater'];
      session.theaterName = sessionJson['theaterName'];
      session.date = sessionJson['date'];
      session.showList = (sessionJson['shows'] as List<Map<String, Object>>).map((showJson) {
        var show = new SessionSearchShow();
        show.canonical = showJson['canonical'];
        show.title = showJson['title'];
        return show;
      });
      return session;
    }).toList();
    return sessionSearchResponse;
  };
}